Delete git/RO directories when not needed, delete old files/dirs in cron

kinow opened the pull request:
Closes #424 
Closes #200 
Closes #279 
Supersedes and closes #344 

## Description
<!--- Describe your changes in detail -->

See #424 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

See #424 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

- [x] Unit tests
- [x] Local docker-compose cluster
- [x] Cron expression (i.e. create files with certain file age, start server, wait for cron/scheduler to run)

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

kinow commented:
Note to self: confirm we are deleting these files when they are no longer necessary, and not too early.

mr-c commented on src/main/java/org/commonwl/view/Scheduler.java:
```diff
@@ -55,4 +81,70 @@ public void removeOldQueuedWorkflowEntries() {
+        List<String> temporaryDirectories = Stream.of(bundleStorage, graphvizStorage, gitStorage)
+                .distinct()
+                .toList();
+        temporaryDirectories.forEach(this::clearDirectory);
```
Can we keep the ROs but delete the old Git checkouts?

kinow replied:
Sure. Do you mean keep all the ROs, or just the ones in the first page of the workflows listing?

I was going to send a screenshot of the first page, but the server appears to be unwell again.

![image](https://user-images.githubusercontent.com/304786/174769370-11db3888-a54c-4e25-a3ff-b1c6aedb354d.png)

I'll leave a message on the Matrix CWLViewer and in some room at Curii to see if someone can kick that server or container, etc.

kinow replied:
Updated the code above to skip bundles :+1:

kinow commented:
Things to think about (from last week's meetings):

- Can we improve compression of bundle files? Choose a different compression level?
- Do we have any workflows with very large files? Or maybe a different approach could be to just use `dd` or use some binary file, putting it in a git repo with a dummy CWL workflow file, and then import it into CWL Viewer, and confirm if the git repo files sizes reflect in the bundle file size.
- The metagenomics workflows have many files, but not sure if there are large files.
- We need to confirm that, having the bundle ZIP, the bundle directory is really not necessary. If so, we can safely delete the bundle directories.
- If the RO bundle ZIP files are not too large, then maybe we can just leave the ZIP files on disk? Or maybe delete the ZIP files using a higher file age limit (1-3 months for the ZIPs? a few hours or days for the cron job that deletes directories, graphs?)

tetron commented:
A data point about the lingering git repositories:

> There’s 1 checked out repo of 1.2GB (250MB of which are gone in the examples/ subdir
> then, there’s multiple copies of ~700MB

So it might also be a good idea to introduce a cap on the size of research object bundles, even if we clean up the git repos, keeping around a bunch of 500-1000 MB RO bundles is a huge disk space hog.

kinow commented:
Note: I spent some time debugging the main flow of importing a workflow, and there are a few things that are not perfect in this PR. Namely, sometimes an exception is triggered as a graph already exists, and that's breaking the process due to the changes here (this exception is old, I've seen it before), and I think I am deleting the `.git` directory, not the git repository. I took notes of what appeared to be wrong while debugging it, and also noting a few other things, and will take care of this next time I have to work on the viewer :+1:

kinow commented:
Rebased, fixed the code I remembered wasn't working as expected, fixed unit tests, and manually testing found no issues.

Waiting for CI to review the code coverage, then either mark as ready for review or add new unit tests. Also need to test the scheduler/cron.

Note to self & reviewers, here's how I tested it:

```bash
# create containers
$ docker-compose up --no-start
# start DB & Jena/SPARQL
$ docker-compose start postgres sparql
# create temporary directory, and separate directories for each service (production uses a single one, but it's handy for testing)
$ tree /tmp/cwlviewer/
/tmp/cwlviewer/
├── bundles
├── git
└── graphviz
$ mkdir -p /tmp/cwlviewer/{bundles,git,graphviz}
# in IntelliJ, start the `CwlViewerApplication` with the following JVM settings:
# -Djava.io.tmpdir=/tmp/cwlviewer/ -DbundleStorage=/tmp/cwlviewer/bundles -DgitStorage=/tmp/cwlviewer/git -DgraphvizStorage=/tmp/cwlviewer/graphviz
# Visit http://localhost:8080/
```

Now import `compile.cwl`. After a few seconds, when it succeeds, you should have:

- `/tmp/cwlviewer/graphviz` with three graphs for `compile.cwl`
- `/tmp/cwlviewer/bundles` with a single ZIP bundle
- `/tmp/cwlviewer/git` empty (the git repository was cloned there, but the `ROBundleService` now deletes the git repository once it has suceeded - no further need for it)
- `/tmp/cwlviewer/` with the three directories above, a couple created by Tomcat/Spring, and no other file or directory

The most important part here is the removal of the Git repositories after they have been successfully used. There are cases where we could have an issue in the middle of the process execution, and the git and bundle directories could be left behind. Unfortunately the only way to remove them is periodically based on some file-age (hence the Scheduler/cron change).

-Bruno

kinow commented:
Just pending some manual testing, and I will try to add a few mocked tests, but looks like the code that's not covered contains mostly exception handling, which is hard to be covered. For the Scheduler tomorrow I will start the server with a shorter cron expression (run every 2 minutes or so), and confirm it's deleting the files correctly.

kinow commented on src/main/java/org/commonwl/view/Scheduler.java:
```diff
@@ -55,4 +79,70 @@ public void removeOldQueuedWorkflowEntries() {
+    /**
+     * Scheduled function to delete old temporary directories.
+     *
+     * <p>Will scan each temporary directory (graphviz, RO, git), searching
```
It is 1., the temporary directory used to create the bundle. We remove it after the bundle was closed (ie the zip file has been created elsewhere).

On my machine I get many empty robundle-$hash dirs  that are not removed. Not really an issue for disk space (maybe inode use? Haha) , so happy to remove this code if you prefer.

kinow commented on src/main/java/org/commonwl/view/researchobject/ROBundleService.java:
```diff
@@ -195,7 +203,7 @@ public Bundle createBundle(Workflow workflow, GitDetails gitInfo) throws IOExcep
                         "merged.cwl", cwlTool.getPackedVersion(rawUrl));
             } catch (CWLValidationException ex) {
-                logger.error("Could not pack workflow when creating Research Object", ex.getMessage());
+                logger.error(String.format("Could not pack workflow when creating Research Object: %s", ex.getMessage()), ex);
```
Bug found with IDE.

kinow commented on src/main/java/org/commonwl/view/researchobject/ROBundleService.java:
```diff
@@ -150,13 +152,19 @@ public Bundle createBundle(Workflow workflow, GitDetails gitInfo) throws IOExcep
+                throw e;
             } finally {
                 gitSemaphore.release(gitInfo.getRepoUrl());
+                org.commonwl.view.util.FileUtils.deleteGitRepository(gitRepo);
```
Remove git folder in the `finally` statement since we won't need it after this point (i.e. the graphs are created, the cwltool has been executed and populated the DB, and now the git artefacts are already in the bundle).

kinow commented:
Something strange with the coverage reports. `FileUtils.java` is now fully covered by `FileUtilsTest.java` :confused:

kinow commented:
Tested the `Scheduler` code, which is executed periodically in the server (akin to cron). I changed in application.properties to, instead of running every mid-night (`cron.clearTmpDir = 0 0 0 * * ?`), to run every five minutes (`cron.clearTmpDir = 5 * * * * ?`).

Then started the docker cluster and imported `compile.cwl` and `make-to-cwl.cwl`. Both succeeded and created some images and two bundles.

```bash
kinow@ranma:/tmp/cwlviewer/graphviz$ ls -lah
total 96K
drwxrwxr-x  2 kinow kinow 4.0K Aug 16 18:34 .
drwxrwxr-x 10 kinow kinow 4.0K Aug 16 18:34 ..
-rw-rw-r--  1 kinow kinow  15K Aug 15 18:35 03e3eb2c-fe6b-4cab-ab8d-db6747bd03bf.png
-rw-rw-r--  1 kinow kinow 7.0K Aug 15 18:35 03e3eb2c-fe6b-4cab-ab8d-db6747bd03bf.svg
-rw-rw-r--  1 kinow kinow  12K Aug 16 18:34 2b3c799c-a9a4-4e15-9cb3-475930324952.png
-rw-rw-r--  1 kinow kinow  22K Aug 16 18:34 8053cbf6-df50-412e-85dc-7465a2303764.png
-rw-rw-r--  1 kinow kinow 7.2K Aug 16 18:34 8053cbf6-df50-412e-85dc-7465a2303764.svg
-rw-rw-r--  1 kinow kinow  19K Aug 16 18:34 da75c0ec-9565-4f2e-9fcf-ef2a453cfda9.png
```

The bundles are never touched as we they are in another directory.

Then I modified the access and modified dates of two files.

```bash
kinow@ranma:/tmp/cwlviewer/graphviz$ date
Tue 16 Aug 2022 18:35:47 NZST
kinow@ranma:/tmp/cwlviewer/graphviz$ touch -a -m -t 202208151835.00 03e3eb2c-fe6b-4cab-ab8d-db6747bd03bf.png
kinow@ranma:/tmp/cwlviewer/graphviz$ touch -a -m -t 202208151835.00 03e3eb2c-fe6b-4cab-ab8d-db6747bd03bf.svg
```

After a few minutes the scheduler runs and deletes the old files.

```bash
kinow@ranma:/tmp/cwlviewer/graphviz$ ls -lah
total 72K
drwxrwxr-x  2 kinow kinow 4.0K Aug 16 18:38 .
drwxrwxr-x 10 kinow kinow 4.0K Aug 16 18:34 ..
-rw-rw-r--  1 kinow kinow  12K Aug 16 18:34 2b3c799c-a9a4-4e15-9cb3-475930324952.png
-rw-rw-r--  1 kinow kinow  22K Aug 16 18:34 8053cbf6-df50-412e-85dc-7465a2303764.png
-rw-rw-r--  1 kinow kinow 7.2K Aug 16 18:34 8053cbf6-df50-412e-85dc-7465a2303764.svg
-rw-rw-r--  1 kinow kinow  19K Aug 16 18:34 da75c0ec-9565-4f2e-9fcf-ef2a453cfda9.png
```

Accessing both workflows in CWL Viewer re-created the images almost instantaneously.

Bundles still OK:

```bash
kinow@ranma:/tmp/cwlviewer/graphviz$ ls -lah /tmp/cwlviewer/bundles/
total 60K
drwxrwxr-x  2 kinow kinow 4.0K Aug 16 18:34 .
drwxrwxr-x 10 kinow kinow 4.0K Aug 16 18:34 ..
-rw-rw-r--  1 kinow kinow  28K Aug 16 18:34 bundle-47e8c47c-91af-44de-a02a-3cb6f20a89a9.zip
-rw-rw-r--  1 kinow kinow  21K Aug 16 18:34 bundle-ac6285ef-e6d5-49df-b988-6105412798e9.zip
```
![Screenshot from 2022-08-16 18-38-31](https://user-images.githubusercontent.com/304786/184814984-07990ca4-b0b7-4089-978e-51b5fbda5051.png)
![Screenshot from 2022-08-16 18-38-08](https://user-images.githubusercontent.com/304786/184814992-56f3288e-cf6b-46bb-86a2-25dc0a208bca.png)

kinow commented:
@mr-c , @tetron ready for review. See previous comments for notes on how to test/how it was tested. One thing I just realized, is that the `Scheduler.java` I included here assumes in production we have distinct folders for bundles, graphs, and git repositories.

The default settings are pointing to the same temporary directory. I do not have access to production, so someone would have to confirm. If we have a single directory, then we can either change it, or I can remove `Scheduler.java` from this PR and move it to another draft PR for later.

Thanks! :vulcan_salute: 
Bruno

mr-c reviewed (approved):
I'm going to merge this, make a pre-release, and ask UniTo to test it

kinow commented:
> I'm going to merge this, make a pre-release, and ask UniTo to test it

Great! I had forgotten about this PR. Hopefully it will help us with disk space issues :crossed_fingers:  Thanks @mr-c !
